commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/serde/ChunkSizesBinaryCodec.java

```diff
@@ -0,0 +1,199 @@
+ *     </tr>
+ *     <tr>
+ *         <td>Count</td>
+ *         <td>4</td>
+ *     </tr>
+ *     <tr>
+ *         <td>Base</td>
+ *         <td>4</td>
+ *     </tr>
+ *     <tr>
+ *         <td>Bytes per value</td>
+ *         <td>1</td>
+ *     </tr>
+ *     <tr>
+ *         <td>Values (without last)</td>
+ *         <td>(Count - 1) * Bytes per value</td>
+ *     </tr>
+ *     <tr>
+ *         <td>Last value</td>
+ *         <td>4</td>
+ *     </tr>
+ * </table>
+ *
+ * <p>When there are zero values, only the count is output.
+ *
+ * <p>When there is one value, only the count and the (last) value is output,
+ * i.e. no base, bytes per value, non-last values.
+ *
+ * <p>The codec accepts only non-negative (>= 0) values.
+ */
+class ChunkSizesBinaryCodec {
+    private static final int COUNT_SIZE = 4;
+    private static final int BASE_SIZE = 4;
+    private static final int BYTES_PER_VALUE_SIZE = 1;
+    private static final int FULL_VALUE_SIZE = 4;
+
+    static byte[] encode(final List<Integer> values) {
+        final int count = values.size();
+        if (count == 0) {
+            final ByteBuffer intBuf = ByteBuffer.allocate(COUNT_SIZE);
+            intBuf.putInt(count);
+            return intBuf.array();
+        }
+
+        final int lastValue = values.get(values.size() - 1);
+
+        if (count == 1) {
+            if (lastValue < 0) {
+                throw new IllegalArgumentException("Values cannot be negative");
+            }
+
+            final ByteBuffer intBuf = ByteBuffer.allocate(COUNT_SIZE + FULL_VALUE_SIZE);
+            intBuf.putInt(count);
+            intBuf.putInt(lastValue);
+            return intBuf.array();
+        }
+
+        final int min = values.stream()
+            // Don't include the last value, it's encoded independently.
+            .limit(values.size() - 1)
+            .mapToInt(Integer::intValue)
+            .min().getAsInt();
+        if (min < 0 || lastValue < 0) {
+            throw new IllegalArgumentException("Values cannot be negative");
+        }
+        final int base = min;
+
+        final byte bytesPerValue = (byte) values.stream()
+            .limit(values.size() - 1)
+            .mapToInt(v -> bytesNeeded(v - base))
+            .max().getAsInt();
+
+        final int bufSize = COUNT_SIZE + BASE_SIZE + BYTES_PER_VALUE_SIZE + ((count - 1) * bytesPerValue)
+            + FULL_VALUE_SIZE;  // last value in full
+        final ByteBuffer buf = ByteBuffer.allocate(bufSize);
+
+        buf.putInt(count);
+        buf.putInt(base);
+        buf.put(bytesPerValue);
+
+        final ByteBuffer intBuf = ByteBuffer.allocate(FULL_VALUE_SIZE);
+        for (int i = 0; i < values.size() - 1; i++) {
+            final int value = values.get(i);
+            final int onBase = value - base;
+            intBuf.rewind();
+            intBuf.putInt(onBase);
+            // The buffer is big-endian.
+            buf.put(intBuf.array(), FULL_VALUE_SIZE - bytesPerValue, bytesPerValue);
```

Comment:
AFAIU offset is known in advance and there is no need to recalculate it in a loop

Reply:
I hope the JIT compiler is smart enough for this, but anyway extracted a constant